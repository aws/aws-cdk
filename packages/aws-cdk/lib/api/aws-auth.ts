export * from '../../../@aws-cdk/toolkit-lib-tmp/lib/api/aws-auth';
