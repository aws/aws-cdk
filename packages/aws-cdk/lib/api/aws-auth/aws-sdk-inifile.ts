import * as AWS from 'aws-sdk';


/**
 * Hack-fix
 *
 * There are a number of issues in the upstream version of SharedIniFileCredentials
 * that need fixing:
 *
 *  1. The upstream aws-sdk contains an incorrect instantiation of an `AWS.STS`
 *     client, which *should* have taken the region from the requested profile
 *     but doesn't. It will use the region from the default profile, which
 *     may not exist, defaulting to `us-east-1` (since we switched to
 *     AWS_STS_REGIONAL_ENDPOINTS=regional, that default is not even allowed anymore
 *     and the absence of a default region will lead to an error).
 *
 *  2. The simple fix is to get the region from the `config` file. profiles
 *     are made up of a combination of `credentials` and `config`, and the region is
 *     generally in `config` with the rest in `credentials`. However, a bug in
 *     `getProfilesFromSharedConfig` overwrites ALL `config` data with `credentials`
 *     data, so we also need to do extra work to fish the `region` out of the config.
 *
 * 3.  The 'credential_source' option is not supported. Meaning credentials
 *     for assume-role cannot be fetched using EC2/ESC metadata.
 *
 * See https://github.com/aws/aws-sdk-js/issues/3418 for all the gory details.
 * See https://github.com/aws/aws-sdk-js/issues/1916 for some more glory details.
 */
export class PatchedSharedIniFileCredentials extends AWS.SharedIniFileCredentials {
  declare private profile: string;
  declare private filename: string;
  declare private disableAssumeRole: boolean;
  declare private options: Record<string, string>;
  declare private roleArn: string;
  declare private httpOptions?: AWS.HTTPOptions;
  declare private tokenCodeFn?: (mfaSerial: string, callback: (err?: Error, token?: string) => void) => void;

  public loadRoleProfile(
    creds: Record<string, Record<string, string>>,
    roleProfile: Record<string, string>,
    callback: (err?: Error, data?: any) => void) {

    // Need to duplicate the whole implementation here -- the function is long and has been written in
    // such a way that there are no small monkey patches possible.

    if (this.disableAssumeRole) {
      throw (AWS as any).util.error(
        new Error('Role assumption profiles are disabled. ' +
                  'Failed to load profile ' + this.profile +
                  ' from ' + creds.filename),
        { code: 'SharedIniFileCredentialsProviderFailure' },
      );
    }

    var self = this;
    var roleArn = roleProfile.role_arn;
    var roleSessionName = roleProfile.role_session_name;
    var externalId = roleProfile.external_id;
    var mfaSerial = roleProfile.mfa_serial;
    var sourceProfile = roleProfile.source_profile;
    var credentialSource = roleProfile.credential_source;

    const credentialError = (AWS as any).util.error(
      new Error(`When using 'role_arn' in profile ('${this.profile}'), you must also configure exactly one of 'source_profile' or 'credential_source'`),
      { code: 'SharedIniFileCredentialsProviderFailure' },
    );

    if (sourceProfile && credentialSource) {
      throw credentialError;
    }

    if (!sourceProfile && !credentialSource) {
      throw credentialError;
    }

    const profiles = loadProfilesProper(this.filename);
    const region = profiles[this.profile]?.region ?? profiles.default?.region ?? 'us-east-1';

    const stsCreds = sourceProfile ? this.sourceProfileCredentials(sourceProfile, creds) : this.credentialSourceCredentials(credentialSource);

    this.roleArn = roleArn;
    var sts = new AWS.STS({
      credentials: stsCreds,
      region,
      httpOptions: this.httpOptions,
    });

    var roleParams: AWS.STS.AssumeRoleRequest = {
      RoleArn: roleArn,
      RoleSessionName: roleSessionName || 'aws-sdk-js-' + Date.now(),
    };

    if (externalId) {
      roleParams.ExternalId = externalId;
    }

    if (mfaSerial && self.tokenCodeFn) {
      roleParams.SerialNumber = mfaSerial;
      self.tokenCodeFn(mfaSerial, function(err, token) {
        if (err) {
          var message;
          if (err instanceof Error) {
            message = err.message;
          } else {
            message = err;
          }
          callback(
            (AWS as any).util.error(
              new Error('Error fetching MFA token: ' + message),
              { code: 'SharedIniFileCredentialsProviderFailure' },
            ));
          return;
        }

        roleParams.TokenCode = token;
        sts.assumeRole(roleParams, callback);
      });
      return;
    }
    sts.assumeRole(roleParams, callback);
  }

  private sourceProfileCredentials(sourceProfile: string, profiles: Record<string, Record<string, string>>) {

    var sourceProfileExistanceTest = profiles[sourceProfile];

    if (typeof sourceProfileExistanceTest !== 'object') {
      throw (AWS as any).util.error(
        new Error('source_profile ' + sourceProfile + ' using profile '
          + this.profile + ' does not exist'),
        { code: 'SharedIniFileCredentialsProviderFailure' },
      );
    }

    return new AWS.SharedIniFileCredentials(
      (AWS as any).util.merge(this.options || {}, {
        profile: sourceProfile,
        preferStaticCredentials: true,
      }),
    );

  }

  // the aws-sdk for js does not support 'credential_source' (https://github.com/aws/aws-sdk-js/issues/1916)
  // so unfortunately we need to implement this ourselves.
  private credentialSourceCredentials(sourceCredential: string) {

    // see https://docs.aws.amazon.com/credref/latest/refdocs/setting-global-credential_source.html
    switch (sourceCredential) {
      case 'Environment': {
        return new AWS.EnvironmentCredentials('AWS');
      }
      case 'Ec2InstanceMetadata': {
        return new AWS.EC2MetadataCredentials();
      }
      case 'EcsContainer': {
        return new AWS.ECSCredentials();
      }
      default: {
        throw new Error(`credential_source ${sourceCredential} in profile ${this.profile} is unsupported. choose one of [Environment, Ec2InstanceMetadata, EcsContainer]`);
      }
    }

  }
}

/**
 * A function to load profiles from disk that MERGES credentials and config instead of overwriting
 *
 * @see https://github.com/aws/aws-sdk-js/blob/5ae5a7d7d24d1000dbc089cc15f8ed2c7b06c542/lib/util.js#L956
 */
function loadProfilesProper(filename: string) {
  const util = (AWS as any).util; // Does exists even though there aren't any typings for it
  const iniLoader = util.iniLoader;
  const profiles: Record<string, Record<string, string>> = {};
  let profilesFromConfig: Record<string, Record<string, string>> = {};
  if (process.env[util.configOptInEnv]) {
    profilesFromConfig = iniLoader.loadFrom({
      isConfig: true,
      filename: process.env[util.sharedConfigFileEnv],
    });
  }
  var profilesFromCreds: Record<string, Record<string, string>> = iniLoader.loadFrom({
    filename: filename ||
      (process.env[util.configOptInEnv] && process.env[util.sharedCredentialsFileEnv]),
  });
  for (const [name, profile] of Object.entries(profilesFromConfig)) {
    profiles[name] = profile;
  }
  for (const [name, profile] of Object.entries(profilesFromCreds)) {
    profiles[name] = {
      ...profiles[name],
      ...profile,
    };
  }
  return profiles;
}
