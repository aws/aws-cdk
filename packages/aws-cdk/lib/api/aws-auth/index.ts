export * from './sdk';
export * from './sdk-provider';
export * from './credentials';