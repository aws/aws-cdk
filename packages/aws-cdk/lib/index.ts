export * from './api';
export * from './plugin';
export { cli } from './cli';
