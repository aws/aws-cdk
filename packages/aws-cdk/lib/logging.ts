import { Writable } from 'stream';
import * as util from 'util';
import * as colors from 'colors/safe';

type StyleFn = (str: string) => string;
const { stdout, stderr } = process;

const logger = (stream: Writable, styles?: StyleFn[]) => (fmt: string, ...args: any[]) => {
  let str = util.format(fmt, ...args);
  if (styles && styles.length) {
    str = styles.reduce((a, style) => style(a), str);
  }
  stream.write(str + '\n');
};

export let logLevel = LogLevel.DEFAULT;

export function setLogLevel(newLogLevel: LogLevel) {
  logLevel = newLogLevel;
}

export function increaseVerbosity() {
  logLevel += 1;
}

const _debug = logger(stderr, [colors.gray]);

export const trace = (fmt: string, ...args: any) => logLevel >= LogLevel.TRACE && _debug(fmt, ...args);
export const debug = (fmt: string, ...args: any[]) => logLevel >= LogLevel.DEBUG && _debug(fmt, ...args);
export const error = logger(stderr, [colors.red]);
export const warning = logger(stderr, [colors.yellow]);
export const success = logger(stderr, [colors.green]);
export const highlight = logger(stderr, [colors.bold]);
export const print = logger(stderr);
export const data = logger(stdout);

export type LoggerFunction = (fmt: string, ...args: any[]) => void;

/**
 * Create a logger output that features a constant prefix string.
 *
 * @param prefixString the prefix string to be appended before any log entry.
 * @param fn   the logger function to be used (typically one of the other functions in this module)
 *
 * @returns a new LoggerFunction.
 */
export function prefix(prefixString: string, fn: LoggerFunction): LoggerFunction {
  return (fmt: string, ...args: any[]) => fn(`%s ${fmt}`, prefixString, ...args);
}

export const enum LogLevel {
  /** Not verbose at all */
  DEFAULT = 0,
  /** Pretty verbose */
  DEBUG = 1,
  /** Extremely verbose */
  TRACE = 2
}
