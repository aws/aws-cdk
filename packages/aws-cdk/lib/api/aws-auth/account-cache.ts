import * as path from 'path';
import * as fs from 'fs-extra';
import { debug } from '../../logging';
import { cdkCacheDir } from '../../util/directories';
import { Account } from './sdk-provider';

/**
 * Disk cache which maps access key IDs to account IDs.
 * Usage:
 *   cache.get(accessKey) => accountId | undefined
 *   cache.put(accessKey, accountId)
 */
export class AccountAccessKeyCache {
  /**
   * Max number of entries in the cache, after which the cache will be reset.
   */
  public static readonly MAX_ENTRIES = 1000;

  private readonly cacheFile: string;

  /**
   * @param filePath Path to the cache file
   */
  constructor(filePath?: string) {
    this.cacheFile = filePath || path.join(cdkCacheDir(), 'accounts_partitions.json');
  }

  /**
   * Tries to fetch the account ID from cache. If it's not in the cache, invokes
   * the resolver function which should retrieve the account ID and return it.
   * Then, it will be stored into disk cache returned.
   *
   * Example:
   *
   *    const accountId = cache.fetch(accessKey, async () => {
   *      return await fetchAccountIdFromSomewhere(accessKey);
   *    });
   *
   * @param accessKeyId
   * @param resolver
   */
  public async fetch<A extends Account>(accessKeyId: string, resolver: () => Promise<A>) {
    // try to get account ID based on this access key ID from disk.
    const cached = await this.get(accessKeyId);
    if (cached) {
      debug(`Retrieved account ID ${cached.accountId} from disk cache`);
      return cached;
    }

    // if it's not in the cache, resolve and put in cache.
    const account = await resolver();
    if (account) {
      await this.put(accessKeyId, account);
    }

    return account;
  }

  /** Get the account ID from an access key or undefined if not in cache */
  public async get(accessKeyId: string): Promise<Account | undefined> {
    const map = await this.loadMap();
    return map[accessKeyId];
  }

  /** Put a mapping betweenn access key and account ID */
  public async put(accessKeyId: string, account: Account) {
    let map = await this.loadMap();

    // nuke cache if it's too big.
    if (Object.keys(map).length >= AccountAccessKeyCache.MAX_ENTRIES) {
      map = { };
    }

    map[accessKeyId] = account;
    await this.saveMap(map);
  }

  private async loadMap(): Promise<{ [accessKeyId: string]: Account }> {
    try {
      return await fs.readJson(this.cacheFile);
    } catch (e) {
      // File doesn't exist or is not readable. This is a cache,
      // pretend we successfully loaded an empty map.
      if (e.code === 'ENOENT' || e.code === 'EACCES') { return {}; }
      // File is not JSON, could be corrupted because of concurrent writes.
      // Again, an empty cache is fine.
      if (e instanceof SyntaxError) { return {}; }
      throw e;
    }
  }

  private async saveMap(map: { [accessKeyId: string]: Account }) {
    try {
      await fs.ensureFile(this.cacheFile);
      await fs.writeJson(this.cacheFile, map, { spaces: 2 });
    } catch (e) {
      // File doesn't exist or file/dir isn't writable. This is a cache,
      // if we can't write it then too bad.
      if (e.code === 'ENOENT' || e.code === 'EACCES' || e.code === 'EROFS') { return; }
      throw e;
    }
  }
}
