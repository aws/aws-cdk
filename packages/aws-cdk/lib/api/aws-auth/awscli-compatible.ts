import * as child_process from 'child_process';
import * as os from 'os';
import * as path from 'path';
import * as util from 'util';
import * as AWS from 'aws-sdk';
import * as fs from 'fs-extra';
import * as promptly from 'promptly';
import { debug } from '../../logging';
import { PatchedSharedIniFileCredentials } from './aws-sdk-inifile';
import { SharedIniFile } from './sdk_ini_file';

/**
 * Behaviors to match AWS CLI
 *
 * See these links:
 *
 * https://docs.aws.amazon.com/cli/latest/topic/config-vars.html
 * https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html
 */
export class AwsCliCompatible {
  /**
   * Build an AWS CLI-compatible credential chain provider
   *
   * This is similar to the default credential provider chain created by the SDK
   * except:
   *
   * 1. Accepts profile argument in the constructor (the SDK must have it prepopulated
   *    in the environment).
   * 2. Conditionally checks EC2 credentials, because checking for EC2
   *    credentials on a non-EC2 machine may lead to long delays (in the best case)
   *    or an exception (in the worst case).
   * 3. Respects $AWS_SHARED_CREDENTIALS_FILE.
   * 4. Respects $AWS_DEFAULT_PROFILE in addition to $AWS_PROFILE.
   */
  public static async credentialChain(options: CredentialChainOptions = {}) {

    // To match AWS CLI behavior, if a profile is explicitly given using --profile,
    // we use that to the exclusion of everything else (note: this does not apply
    // to AWS_PROFILE, environment credentials still take precedence over AWS_PROFILE)
    if (options.profile) {
      await forceSdkToReadConfigIfPresent();
      const theProfile = options.profile;
      return new AWS.CredentialProviderChain([
        () => profileCredentials(theProfile),
        () => new AWS.ProcessCredentials({ profile: theProfile }),
      ]);
    }

    const implicitProfile = process.env.AWS_PROFILE || process.env.AWS_DEFAULT_PROFILE || 'default';

    const sources = [
      () => new AWS.EnvironmentCredentials('AWS'),
      () => new AWS.EnvironmentCredentials('AMAZON'),
    ];

    if (await fs.pathExists(credentialsFileName())) {
      // Force reading the `config` file if it exists by setting the appropriate
      // environment variable.
      await forceSdkToReadConfigIfPresent();
      sources.push(() => profileCredentials(implicitProfile));
      sources.push(() => new AWS.ProcessCredentials({ profile: implicitProfile }));
    }

    if (options.containerCreds ?? hasEcsCredentials()) {
      sources.push(() => new AWS.ECSCredentials());
    } else if (hasWebIdentityCredentials()) {
      // else if: we have found WebIdentityCredentials as provided by EKS ServiceAccounts
      sources.push(() => new AWS.TokenFileWebIdentityCredentials());
    } else if (options.ec2instance ?? await isEc2Instance()) {
      // else if: don't get EC2 creds if we should have gotten ECS or EKS creds
      // ECS and EKS instances also run on EC2 boxes but the creds represent something different.
      // Same behavior as upstream code.
      sources.push(() => new AWS.EC2MetadataCredentials());
    }

    return new AWS.CredentialProviderChain(sources);

    function profileCredentials(profileName: string) {
      return new PatchedSharedIniFileCredentials({
        profile: profileName,
        filename: credentialsFileName(),
        httpOptions: options.httpOptions,
        tokenCodeFn,
      });
    }
  }

  /**
   * Return the default region in a CLI-compatible way
   *
   * Mostly copied from node_loader.js, but with the following differences to make it
   * AWS CLI compatible:
   *
   * 1. Takes a profile name as an argument (instead of forcing it to be taken from $AWS_PROFILE).
   *    This requires having made a copy of the SDK's `SharedIniFile` (the original
   *    does not take an argument).
   * 2. $AWS_DEFAULT_PROFILE and $AWS_DEFAULT_REGION are also respected.
   *
   * Lambda and CodeBuild set the $AWS_REGION variable.
   */
  public static async region(options: RegionOptions = {}): Promise<string> {
    const profile = options.profile || process.env.AWS_PROFILE || process.env.AWS_DEFAULT_PROFILE || 'default';

    // Defaults inside constructor
    const toCheck = [
      { filename: credentialsFileName(), profile },
      { isConfig: true, filename: configFileName(), profile },
      { isConfig: true, filename: configFileName(), profile: 'default' },
    ];

    let region = process.env.AWS_REGION || process.env.AMAZON_REGION ||
      process.env.AWS_DEFAULT_REGION || process.env.AMAZON_DEFAULT_REGION;

    while (!region && toCheck.length > 0) {
      const opts = toCheck.shift()!;
      if (await fs.pathExists(opts.filename)) {
        const configFile = new SharedIniFile(opts);
        const section = await configFile.getProfile(opts.profile);
        region = section?.region;
      }
    }

    if (!region && (options.ec2instance ?? await isEc2Instance())) {
      debug('Looking up AWS region in the EC2 Instance Metadata Service (IMDS).');
      const imdsOptions = {
        httpOptions: { timeout: 1000, connectTimeout: 1000 }, maxRetries: 2,
      };
      const metadataService = new AWS.MetadataService(imdsOptions);

      let token;
      try {
        token = await getImdsV2Token(metadataService);
      } catch (e) {
        debug(`No IMDSv2 token: ${e}`);
      }

      try {
        region = await getRegionFromImds(metadataService, token);
        debug(`AWS region from IMDS: ${region}`);
      } catch (e) {
        debug(`Unable to retrieve AWS region from IMDS: ${e}`);
      }
    }

    if (!region) {
      const usedProfile = !profile ? '' : ` (profile: "${profile}")`;
      region = 'us-east-1'; // This is what the AWS CLI does
      debug(`Unable to determine AWS region from environment or AWS configuration${usedProfile}, defaulting to '${region}'`);
    }

    return region;
  }
}

/**
 * Return whether it looks like we'll have ECS credentials available
 */
function hasEcsCredentials(): boolean {
  return (AWS.ECSCredentials.prototype as any).isConfiguredForEcsCredentials();
}

/**
 * Return whether it looks like we'll have WebIdentityCredentials (that's what EKS uses) available
 * No check like hasEcsCredentials available, so have to implement our own.
 * @see https://github.com/aws/aws-sdk-js/blob/3ccfd94da07234ae87037f55c138392f38b6881d/lib/credentials/token_file_web_identity_credentials.js#L59
 */
function hasWebIdentityCredentials(): boolean {
  return Boolean(process.env.AWS_ROLE_ARN && process.env.AWS_WEB_IDENTITY_TOKEN_FILE);
}

/**
 * Return whether we're on an EC2 instance
 */
async function isEc2Instance() {
  if (isEc2InstanceCache === undefined) {
    debug("Determining if we're on an EC2 instance.");
    let instance = false;
    if (process.platform === 'win32') {
      // https://docs.aws.amazon.com/AWSEC2/latest/WindowsGuide/identify_ec2_instances.html
      const result = await util.promisify(child_process.exec)('wmic path win32_computersystemproduct get uuid', { encoding: 'utf-8' });
      // output looks like
      //  UUID
      //  EC2AE145-D1DC-13B2-94ED-01234ABCDEF
      const lines = result.stdout.toString().split('\n');
      instance = lines.some(x => matchesRegex(/^ec2/i, x));
    } else {
      // https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/identify_ec2_instances.html
      const files: Array<[string, RegExp]> = [
        // This recognizes the Xen hypervisor based instances (pre-5th gen)
        ['/sys/hypervisor/uuid', /^ec2/i],

        // This recognizes the new Hypervisor (5th-gen instances and higher)
        // Can't use the advertised file '/sys/devices/virtual/dmi/id/product_uuid' because it requires root to read.
        // Instead, sys_vendor contains something like 'Amazon EC2'.
        ['/sys/devices/virtual/dmi/id/sys_vendor', /ec2/i],
      ];
      for (const [file, re] of files) {
        if (matchesRegex(re, readIfPossible(file))) {
          instance = true;
          break;
        }
      }
    }
    debug(instance ? 'Looks like an EC2 instance.' : 'Does not look like an EC2 instance.');
    isEc2InstanceCache = instance;
  }
  return isEc2InstanceCache;
}


let isEc2InstanceCache: boolean | undefined = undefined;

/**
 * Attempts to get a Instance Metadata Service V2 token
 */
async function getImdsV2Token(metadataService: AWS.MetadataService): Promise<string> {
  debug('Attempting to retrieve an IMDSv2 token.');
  return new Promise((resolve, reject) => {
    metadataService.request(
      '/latest/api/token',
      {
        method: 'PUT',
        headers: { 'x-aws-ec2-metadata-token-ttl-seconds': '60' },
      },
      (err: AWS.AWSError, token: string | undefined) => {
        if (err) {
          reject(err);
        } else if (!token) {
          reject(new Error('IMDS did not return a token.'));
        } else {
          resolve(token);
        }
      });
  });
}

/**
 * Attempts to get the region from the Instance Metadata Service
 */
async function getRegionFromImds(metadataService: AWS.MetadataService, token: string | undefined): Promise<string> {
  debug('Retrieving the AWS region from the IMDS.');
  let options: { method?: string | undefined; headers?: { [key: string]: string; } | undefined; } = {};
  if (token) {
    options = { headers: { 'x-aws-ec2-metadata-token': token } };
  }
  return new Promise((resolve, reject) => {
    metadataService.request(
      '/latest/dynamic/instance-identity/document',
      options,
      (err: AWS.AWSError, instanceIdentityDocument: string | undefined) => {
        if (err) {
          reject(err);
        } else if (!instanceIdentityDocument) {
          reject(new Error('IMDS did not return an Instance Identity Document.'));
        } else {
          try {
            resolve(JSON.parse(instanceIdentityDocument).region);
          } catch (e) {
            reject(e);
          }
        }
      });
  });
}

function homeDir() {
  return process.env.HOME || process.env.USERPROFILE
    || (process.env.HOMEPATH ? ((process.env.HOMEDRIVE || 'C:/') + process.env.HOMEPATH) : null) || os.homedir();
}

function credentialsFileName() {
  return process.env.AWS_SHARED_CREDENTIALS_FILE || path.join(homeDir(), '.aws', 'credentials');
}

function configFileName() {
  return process.env.AWS_CONFIG_FILE || path.join(homeDir(), '.aws', 'config');
}

/**
 * Force the JS SDK to honor the ~/.aws/config file (and various settings therein)
 *
 * For example, there is just *NO* way to do AssumeRole credentials as long as AWS_SDK_LOAD_CONFIG is not set,
 * or read credentials from that file.
 *
 * The SDK crashes if the variable is set but the file does not exist, so conditionally set it.
 */
async function forceSdkToReadConfigIfPresent() {
  if (await fs.pathExists(configFileName())) {
    process.env.AWS_SDK_LOAD_CONFIG = '1';
  }
}

function matchesRegex(re: RegExp, s: string | undefined) {
  return s !== undefined && re.exec(s) !== null;
}

/**
 * Read a file if it exists, or return undefined
 *
 * Not async because it is used in the constructor
 */
function readIfPossible(filename: string): string | undefined {
  try {
    if (!fs.pathExistsSync(filename)) { return undefined; }
    return fs.readFileSync(filename, { encoding: 'utf-8' });
  } catch (e) {
    debug(e);
    return undefined;
  }
}

export interface CredentialChainOptions {
  readonly profile?: string;
  readonly ec2instance?: boolean;
  readonly containerCreds?: boolean;
  readonly httpOptions?: AWS.HTTPOptions;
}

export interface RegionOptions {
  readonly profile?: string;
  readonly ec2instance?: boolean;
}

/**
 * Ask user for MFA token for given serial
 *
 * Result is send to callback function for SDK to authorize the request
 */
async function tokenCodeFn(serialArn: string, cb: (err?: Error, token?: string) => void): Promise<void> {
  debug('Require MFA token for serial ARN', serialArn);
  try {
    const token: string = await promptly.prompt(`MFA token for ${serialArn}: `, {
      trim: true,
      default: '',
    });
    debug('Successfully got MFA token from user');
    cb(undefined, token);
  } catch (err) {
    debug('Failed to get MFA token', err);
    cb(err);
  }
}

