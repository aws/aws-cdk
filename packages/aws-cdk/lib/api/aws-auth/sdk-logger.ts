import { inspect } from 'util';
import { Logger } from '@smithy/types';
import { trace } from '../../logging';

export class SdkToCliLogger implements Logger {
  public trace(..._content: any[]) {
    // This is too much detail for our logs
    // trace('[SDK trace] %s', fmtContent(content));
  }

  public debug(..._content: any[]) {
    // This is too much detail for our logs
    // trace('[SDK debug] %s', fmtContent(content));
  }

  /**
   * Info is called mostly (exclusively?) for successful API calls
   *
   * Payload:
   *
   * (Note the input contains entire CFN templates, for example)
   *
   * ```
   * {
   *   clientName: 'S3Client',
   *   commandName: 'GetBucketLocationCommand',
   *   input: {
   *     Bucket: '.....',
   *     ExpectedBucketOwner: undefined
   *   },
   *   output: { LocationConstraint: 'eu-central-1' },
   *   metadata: {
   *     httpStatusCode: 200,
   *     requestId: '....',
   *     extendedRequestId: '...',
   *     cfId: undefined,
   *     attempts: 1,
   *     totalRetryDelay: 0
   *   }
   * }
   * ```
   */
  public info(...content: any[]) {
    trace('[sdk info] %s', formatSdkLoggerContent(content));
  }

  public warn(...content: any[]) {
    trace('[sdk warn] %s', formatSdkLoggerContent(content));
  }

  /**
   * Error is called mostly (exclusively?) for failing API calls
   *
   * Payload (input would be the entire API call arguments).
   *
   * ```
   * {
   *   clientName: 'STSClient',
   *   commandName: 'GetCallerIdentityCommand',
   *   input: {},
   *   error: AggregateError [ECONNREFUSED]:
   *       at internalConnectMultiple (node:net:1121:18)
   *       at afterConnectMultiple (node:net:1688:7) {
   *     code: 'ECONNREFUSED',
   *     '$metadata': { attempts: 3, totalRetryDelay: 600 },
   *     [errors]: [ [Error], [Error] ]
   *   },
   *   metadata: { attempts: 3, totalRetryDelay: 600 }
   * }
   * ```
   */
  public error(...content: any[]) {
    trace('[sdk error] %s', formatSdkLoggerContent(content));
  }
}

/**
 * This can be anything.
 *
 * For debug, it seems to be mostly strings.
 * For info, it seems to be objects.
 *
 * Stringify and join without separator.
 */
export function formatSdkLoggerContent(content: any[]) {
  if (content.length === 1) {
    const apiFmt = formatApiCall(content[0]);
    if (apiFmt) {
      return apiFmt;
    }
  }
  return content.map((x) => typeof x === 'string' ? x : inspect(x)).join('');
}

function formatApiCall(content: any): string | undefined {
  if (!isSdkApiCallSuccess(content) && !isSdkApiCallError(content)) {
    return undefined;
  }

  const service = content.clientName.replace(/Client$/, '');
  const api = content.commandName.replace(/Command$/, '');

  const parts = [];
  if ((content.metadata?.attempts ?? 0) > 1) {
    parts.push(`[${content.metadata?.attempts} attempts, ${content.metadata?.totalRetryDelay}ms retry]`);
  }

  parts.push(`${service}.${api}(${JSON.stringify(content.input)})`);

  if (isSdkApiCallSuccess(content)) {
    parts.push('-> OK');
  } else {
    parts.push(`-> ${content.error}`);
  }

  return parts.join(' ');
}

interface SdkApiCallBase {
  clientName: string;
  commandName: string;
  input: Record<string, unknown>;
  metadata?: {
    httpStatusCode?: number;
    requestId?: string;
    extendedRequestId?: string;
    cfId?: string;
    attempts?: number;
    totalRetryDelay?: number;
  };
}

type SdkApiCallSuccess = SdkApiCallBase & { output: Record<string, unknown> };
type SdkApiCallError = SdkApiCallBase & { error: Error };

function isSdkApiCallSuccess(x: any): x is SdkApiCallSuccess {
  return x && typeof x === 'object' && x.commandName && x.output;
}

function isSdkApiCallError(x: any): x is SdkApiCallError {
  return x && typeof x === 'object' && x.commandName && x.error;
}