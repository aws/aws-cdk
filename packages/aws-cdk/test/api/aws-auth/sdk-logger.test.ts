import { formatSdkLoggerContent, SdkToCliLogger } from '../../../lib/api/aws-auth/sdk-logger';
import * as logging from '../../../lib/logging';

describe(SdkToCliLogger, () => {
  const logger = new SdkToCliLogger();
  const trace = jest.spyOn(logging, 'trace');

  beforeEach(() => {
    trace.mockReset();
  });

  test.each(['trace', 'debug'] as Array<keyof SdkToCliLogger>)('%s method does not call trace', (meth) => {
    logger[meth]('test');
    expect(trace).not.toHaveBeenCalled();
  });

  test.each(['info', 'warn', 'error'] as Array<keyof SdkToCliLogger>)('%s method logs to trace', (meth) => {
    logger[meth]('test');
    expect(trace).toHaveBeenCalled();
  });
});

const SUCCESS_CALL = {
  clientName: 'S3Client',
  commandName: 'GetBucketLocationCommand',
  input: {
    Bucket: '.....',
    ExpectedBucketOwner: undefined,
  },
  output: { LocationConstraint: 'eu-central-1' },
  metadata: {
    httpStatusCode: 200,
    requestId: '....',
    extendedRequestId: '...',
    cfId: undefined,
    attempts: 2,
    totalRetryDelay: 30,
  },
};

const ERROR_CALL = {
  clientName: 'S3Client',
  commandName: 'GetBucketLocationCommand',
  input: {
    Bucket: '.....',
    ExpectedBucketOwner: undefined,
  },
  error: new Error('it failed'),
  metadata: {
    httpStatusCode: 200,
    attempts: 2,
    totalRetryDelay: 30,
  },
};

test('formatting a successful SDK call looks broadly reasonable', () => {
  const output = formatSdkLoggerContent([SUCCESS_CALL]);
  expect(output).toMatchSnapshot();
});

test('formatting a failing SDK call looks broadly reasonable', () => {
  const output = formatSdkLoggerContent([ERROR_CALL]);
  expect(output).toMatchSnapshot();
});

test('formatting a successful SDK call includes attempts and retries if greater than 1', () => {
  const output = formatSdkLoggerContent([SUCCESS_CALL]);
  expect(output).toContain('2 attempts');
  expect(output).toContain('30ms');
});

test('formatting a failing SDK call includes attempts and retries if greater than 1', () => {
  const output = formatSdkLoggerContent([ERROR_CALL]);
  expect(output).toContain('2 attempts');
  expect(output).toContain('30ms');
});

test('formatting a failing SDK call includes the error', () => {
  const output = formatSdkLoggerContent([ERROR_CALL]);
  expect(output).toContain('it failed');
});