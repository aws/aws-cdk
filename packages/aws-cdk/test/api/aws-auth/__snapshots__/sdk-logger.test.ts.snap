// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatting a failing SDK call looks broadly reasonable 1`] = `"[2 attempts, 30ms retry] S3.GetBucketLocation({"Bucket":"....."}) -> Error: it failed"`;

exports[`formatting a successful SDK call looks broadly reasonable 1`] = `"[2 attempts, 30ms retry] S3.GetBucketLocation({"Bucket":"....."}) -> OK"`;
